[X86] Combine VPERMV3 to VPERMV for i8/i16

phoebewang opened the pull request:
VPERM[I,T]2[B,W] are 3 uops on Skylake and Icelake so we try to use VPERMV.

RKSimon reviewed (commented):
It might be better to handle this inside combineTargetShufle

phoebewang committed:
[X86] Combine VPERMV3 to VPERMV for i8/i16

VPERM[I,T]2[B,W] are 3 uops on Skylake and Icelake so we try to use VPERMV if possible.

phoebewang commented:
> It might be better to handle this inside combineTargetShufle

Good point! Done.

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -41273,6 +41273,33 @@ static SDValue combineTargetShuffle(SDValue N, const SDLoc &DL,
+    // VPERM[I,T]2[B,W] are 3 uops on Skylake and Icelake so we try to use
+    // VPERMV.
+    if (VT.is512BitVector() || (VT.is256BitVector() && !Subtarget.hasEVEX512()))
+      return SDValue();
```
Not sure, but it might be better to check the NVT type instead of VT?

phoebewang replied:
The drawback is we need to construct NVT before we know if it's necessary. Let me know if you still think checking NVT is better.

RKSimon replied:
MVTs are cheap to construct - I'd say go for it.

phoebewang replied:
Done.

RKSimon commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -41273,6 +41273,33 @@ static SDValue combineTargetShuffle(SDValue N, const SDLoc &DL,
+      return SDValue();
+    MVT ElementVT = VT.getVectorElementType();
+    if (ElementVT != MVT::i8 && ElementVT != MVT::i16)
+      return SDValue();
```
Although VPERM2B/W are slower, do would not still avoid an extract subvector by widening D/Q/PS/PD cases?

phoebewang replied:
But we are also trying to avoid generating 512-bit instructions. I think I can put it in a separate patch in case it has negative effects.

goldsteinn replied:
Can we try to scale the `i8`/`i16` cases?

phoebewang replied:
Do you mean to turn them into `i32`? I think it relies on specific mask patterns, I'm not sure if we have done that or not. Maybe worth a try if not.

goldsteinn replied:
It does rely on specific mask patterns (see `scaleShuffleElements`).

phoebewang replied:
> Although VPERM2B/W are slower, do would not still avoid an extract subvector by widening D/Q/PS/PD cases?

#97206

goldsteinn commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -41273,6 +41273,33 @@ static SDValue combineTargetShuffle(SDValue N, const SDLoc &DL,
+    if (V1.getOpcode() == ISD::EXTRACT_SUBVECTOR &&
+        V1.getConstantOperandVal(1) == 0 &&
+        V2.getOpcode() == ISD::EXTRACT_SUBVECTOR &&
+        V2.getConstantOperandVal(1) == VT.getVectorNumElements() &&
```
Should we also handle the inverse (`V2.getConstantOperandVal(1) == 0`)?

phoebewang replied:
When we built `VECTOR_SHUFFLE` with these two `EXTRACT_SUBVECTOR`, the order is `V1` from 0 and `V2` from the half. So unless there are intermediate combines switch them (which I haven't found), we don't need to handle the inverse one.

RKSimon reviewed (approved):
LGTM
